### ISSUE TITLE
[REQUEST]  Button remap for audi A6 C7

### ISSUE THREAD

**Issue by zrnec** [jilleb/mib2-toolbox#283]
Could it be possible to remap any of media buttons (TEL for example) to open carplay directly, to not go into menu and scroll every time.
Or if anyone could give me pointers which files in firmware are used to manage these buttons.
FW: MHI2_ER_AU57X_K3663
![image](https://github.com/user-attachments/assets/44401df9-d3c9-4a70-8ee5-fae58d46cb00)

**Comment by mrfixpl** [comment]
I've seen https://mqbtools.nl/buttons/ for VW/Seat/Skoda and was hoping that Audi will have the same config file. Sadly, it does not... Would love to remap one of the buttons (or for example long press action of TEL) to ASI as well.

**Comment by zrnec** [comment]
> I've seen https://mqbtools.nl/buttons/ for VW/Seat/Skoda and was hoping that Audi will have the same config file. Sadly, it does not... Would love to remap one of the buttons (or for example long press action of TEL) to ASI as well.

Yep i tried to dump system files using MIB and i coudnl't find any similar/same files as VW uses, i tried finding any files that could handle these pressed but due to lack of my knowledge i couldn't find anything

**Comment by mrfixpl** [comment]
OK, so it looks like Audi, Porsche, Bentley have hardcoded button DSI signals to the keyActions inside the `LSD.JXE`.

Not only the "regular" button presses are defined there, but also `ComboKey` actions - this is mapping two DSI signals with a keyAction like: REM, GEM, HMI_EM, screenshot, reboot, log dump, and emulating buttons that are brand specific.

Some actions are even conditioned with `!this.framework.isPBuild()` to not work if HMI is marked as Production. Or `this.framework.isPorsche()` to work only on specific brand.

**Comment by zrnec** [comment]
> OK, so it looks like Audi, Porsche, Bentley have hardcoded button DSI signals to the keyActions inside the `LSD.JXE`.
> 
> Not only the "regular" button presses are defined there, but also `ComboKey` actions - this is mapping two DSI signals with a keyAction like: REM, GEM, HMI_EM, screenshot, reboot, log dump, and emulating buttons that are brand specific.
> 
> Some actions are even conditioned with `!this.framework.isPBuild()` to not work if HMI is marked as Production. Or `this.framework.isPorsche()` to work only on specific brand.

Sorry i tried to look into some classes inside dump i made with MIB and i can't find anything, based on what you said do you think its possible to make patches to create custom buttons or not? Thanks

**Comment by mrfixpl** [comment]
After further investigation I can confirm that button remapping on Audi/Porsche/Bentley MHI2 is possible via custom java class (same as the NavIgnore patch for dual navigation screen).

Unfortunately, as far as I can see, AU/PO/BE MHI2 LSD.jxe does not have support for APP button that would lead to CarPlay/AndroidAuto.
